### ISSUE TITLE
Conv3d fails with bigger batch size

### ISSUE THREAD

**Issue by alexeib** [pytorch/pytorch#22496]
```
>>> conv = nn.Conv3d(3, 64, kernel_size=(1, 7, 7), stride=(1, 2, 2), padding=(0, 3, 3), bias=False).cuda()
>>> x = torch.FloatTensor(48, 3, 64, 224, 224).cuda()
>>> conv(x).max()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/home/abaevski/.conda/envs/fairseq-20190509/lib/python3.6/site-packages/torch/nn/modules/module.py", line 493, in __call__
    result = self.forward(*input, **kwargs)
  File "/private/home/abaevski/.conda/envs/fairseq-20190509/lib/python3.6/site-packages/torch/nn/modules/conv.py", line 476, in forward
    self.padding, self.dilation, self.groups)
RuntimeError: cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input.
```
works with batchsize up to 41 and starts failing at 42+

cc @ngimel @csarofeen @ptrblck

**Comment by SsnL** [comment]
system info?

**Comment by alexeib** [comment]
1) cuda/9.2   2) cudnn/v7.3-cuda.9.2   3) NCCL/2.2.13-1-cuda.9.2

**Comment by alexeib** [comment]
also broken on 2) cuda/10.0   3) cudnn/v7.4-cuda.10.0   4) NCCL/2.4.2-1-cuda.10.0

**Comment by ngimel** [comment]
Reproduced on the latest cudnn version (7.6), we are looking into it.

**Comment by ngimel** [comment]
Actually, this is a documented restriction - cudnn does not support tensors with more than 2**31 elements, and the output tensor for batch size >=42 has more than 2**31 elements 
https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnSetTensorNdDescriptor

**Comment by bionick87** [comment]
Hi ngimel,
 
Thanks. 

Is it possible to contact the cuDNN nvidia and  do an upgrade? I believe that in the near future the use of  volumes (medical) or video applications will require much more huge input, especially in high definition.
I am currently working with medical volumes.

**Comment by ngimel** [comment]
We can work around this in pytorch, by splitting a single convolution into a few calls, each with less than 2**31 elements. One of our team members will work on this.

**Comment by bionick87** [comment]
Hi ngimel,

This is really good, we really need it, we work with medical data. Our volumes are really huge.

I also observed that a process of volumetric upsamilng tends to occupy a lot of global memory, certainly for us in the medical field it would be necessary to have the possibility of an optimisation of the 3d kernels also from a memory location point of view.

**Comment by zasdfgbnm** [comment]
@ngimel I am self assigning this

**Comment by zasdfgbnm** [comment]
Only partially fixed

**Comment by zasdfgbnm** [comment]
Fixed by: https://github.com/pytorch/pytorch/pull/31379 https://github.com/pytorch/pytorch/pull/31383 https://github.com/pytorch/pytorch/pull/31510 https://github.com/pytorch/pytorch/pull/31524 https://github.com/pytorch/pytorch/pull/31889

When all of them get merged, this can close.

**Comment by michaelklachko** [comment]
Will there be a performance impact from this batch splitting solution? In the original issue https://github.com/pytorch/pytorch/issues/4107 the whole point was to use larger batch sizes (worked fine with smaller batches).

**Comment by zasdfgbnm** [comment]
@michaelklachko It will have comparable performance as you split on the batch dimension manually.

**Comment by michaelklachko** [comment]
To clarify - currently if I want to use batch size 512 to improve my GPU utilization, I will get this CUDNN error. Using batch size 256 is fine. You're saying that you will let me specify batch size 512, but internally you will be splitting it into 2x256, so my question, will the performance be better than me specifying batch size 256? By performance I mean GPU utilization and ultimately wall clock time to converge.

**Comment by zasdfgbnm** [comment]
@michaelklachko 

Generally, the overall GPU utilization will improve if you use a larger batch size. With the batch splitting approach, there is nothing preventing this from happening. Convolution itself benefits little from this approach. But unless your model is just a single convolution layer, using larger batch size could make all your intermediate results being larger so that the rest of your model could benefit from it.

But the GPU utilization and ultimately wall clock time to converge could be dependent on models. So I would say generally yes, but please try it out to see if there is speedup in your case.

**Comment by michaelklachko** [comment]
I just looked at https://github.com/pytorch/pytorch/pull/31379/commits/25cfab2c1f62dabefa1ea0551fef1a262c9b49a7 
Will the loop at line 927 be executed sequentially or in parallel on GPU?

**Comment by ngimel** [comment]
It will be executed sequentially, when operating on 2**31 element tensors there's usually nothing to be gained from parallel execution.

**Comment by michaelklachko** [comment]
If my GPU can fit batch size 512 in memory, then it should be better (faster) to have it process two batches of 256 in parallel than sequentially. What do you mean there's nothing to be gained from that?

**Comment by SsnL** [comment]
@michaelklachko Memory is not the only limit of a GPU. Your GPU has computing cores that run at limited frequencies. At that large of an input (2^31 elements), usually a single op takes all cores at full speed so you can't run things in parallel anyways.

**Comment by ngimel** [comment]
@michaelklachko if your complaint is that with batch 256 and tensor with ~2**31 elements 3d convolution is very slow, that's possible, but that's due to inefficient implementation in cudnn that would unlikely be better even if cudnn supported more than 2**31 element tensors. If you have benchmarks demonstrating such poor performance on important use cases, I encourage you to file a separate issue or bug with nvidia about cudnn performance, but pytorch cannot do anything about that until cudnn provides efficient implementation.

**Comment by msinclair-hf** [comment]
I have encountered this same error in the backward pass of a simple 2 convolution layer model. This occurs when using an input tensor of size [1, 1, 256, 256, 256] (or larger image dimensions), as show in the code snippet below.

Any of the following changes will resolve the error:
1) replace the dilated convolution layer with a regular convolution layer
2) swap the order of the two convolution layers (i.e. place the dilated convolution layer first)
3) for the second convolution layer which has dilation=2, change from padding=2 to padding=1 (which requires reduction of osize=isize-2)
4) use CPU instead of GPU (by removing .cuda() throughout)
5) reduce isize to any number <=255

While I have created a work around by simply replacing the dilated convolution in a U-net with a regular convolution (only required at the first scale), this behaviour suggests further investigation may be required. The tensor size is also only 256^3 = 2^24, i.e. < 2^31, thus appearing not to violate the cudnn limit of 2^31. 


System information: 
OS: Centos 7.7
GPU: Tesla V100
PyTorch version: 1.3.1
CUDA Version: 10.1
Driver version: 418.87.01
cudnn version: 7.6.4

```
import torch
from torch import nn
from torchsummary import summary

class Simple2Conv(nn.Module):
    def __init__(self):
        super(Simple2Conv, self).__init__()
        self.layer_1 = nn.Conv3d(1, 1, 3, padding=1)
        self.layer_2 = nn.Conv3d(1, 1, 3, padding=2, dilation=2)

    def forward(self, x):
        layer_1 = self.layer_1(x)        
        layer_2 = self.layer_2(layer_1)  
        print(layer_2.shape)
        return layer_2

model = Simple2Conv().cuda()

isize = 256
osize = isize 
x = torch.FloatTensor(1,1,isize,isize,isize).cuda()

output = model(x)
criterion = nn.CrossEntropyLoss().cuda()
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)
y = torch.LongTensor(1,osize ,osize ,osize).cuda()

loss = criterion(output, y)
optimizer.zero_grad()
loss.backward()
```

**Comment by ngimel** [comment]
@msinclair-hf it looks like all your tensors are less then 2**31 elements, so this is a different issue. Please open a separate issue.

**Comment by msinclair-hf** [comment]
@ngimel sure, I have opened issue #32035 and have elaborated further on the problem there.

**Comment by michaelklachko** [comment]
> It will be executed sequentially, when operating on 2**31 element tensors there's usually nothing to be gained from parallel execution.

Ok, makes sense. Do you know how Tensorflow handles this case?

**Comment by ngimel** [comment]
No I don't, sorry, but if they are calling into cudnn, they are most likely doing the same thing, because it is a cudnn limitation. Even though technically you can use different streams and run smaller convolutions "concurrently", in reality they will still be serialized because all the device resources will be used by a single call.

**Comment by ekhahniii** [comment]
Experienced this issue for nelement > 2**31. Picked up recent [changes](https://github.com/pytorch/pytorch/issues/22496#issuecomment-571266749) on master from @zasdfgbnm. Issue is resolved in my use case.

**Comment by ngimel** [comment]
CLosing, original issue is solved.
